### PR TITLE
Fixing voter history based election in FlexibleVoteCounter

### PR DESCRIPTION
**Pull request by abhinav04sharma**
Fixing a bunch of things in FlexibleVoteCounter:
 - When enabled, considering majority in candidate region in voter history code path
 - New gflag to control how long we try to win using pessimistic quorum when there is discontinuation in terms
 - Fixed some bugs and restructured some code along the way

### CODE REVIEW

File: src/kudu/consensus/leader_election.cc [hunk 1]
```diff
@@ -556,32 +563,30 @@ FlexibleVoteCounter::IsMajoritySatisfiedInMajorityOfRegions() const {
...
 
   bool quorum_satisfied = true;
   bool quorum_satisfaction_possible = true;
 
-  for (const std::pair<bool, bool>& result : results) {
+  for (const std::pair<bool, bool> &result : results) {
     quorum_satisfied = quorum_satisfied && result.first;
     quorum_satisfaction_possible =
-        quorum_satisfaction_possible || result.second;
+        quorum_satisfaction_possible && result.second;
```
Comment (abhinav04sharma) [inline src/kudu/consensus/leader_election.cc:576]: Without this change quorum_satisfaction_possible will always be true.

File: src/kudu/consensus/leader_election.cc [hunk 2]
```diff
@@ -802,27 +807,15 @@ void FlexibleVoteCounter::AppendPotentialLeaderUUID(
   bool quorum_satisfied = true;
   bool quorum_satisfaction_possible = true;
 
-  if (FLAGS_srd_strict_leader_election_quorum) {
```
Comment (abhinav04sharma) [inline src/kudu/consensus/leader_election.cc:805]: Moving this check to one place in `AreAllMajoritiesSatisfied()`. However, this means that a potential leader is now only based on intersection with the previous leader, we don't take into consideration special majorities like `FLAGS_srd_strict_leader_election_quorum` and `FLAGS_include_candidate_region` etc. which I think is fine since we're only making the case more pessimistic.

File: src/kudu/consensus/leader_election.cc [hunk 3]
```diff
@@ -960,24 +955,16 @@ std::pair<bool, bool> FlexibleVoteCounter::ComputeElectionResultFromVotingHistor
...
-          return std::make_pair<>(
-              result.first && majority_result.first,
-              result.second && majority_result.second);
-        }
-        return result;
+            << JoinStringsIterator(r.potential_leader_regions.begin(),
+                                   r.potential_leader_regions.end(), ", ");
+
+        return AreMajoritiesSatisfied(r.potential_leader_regions,
+                                      candidate_region);
```
Comment (abhinav04sharma) [inline src/kudu/consensus/leader_election.cc:953]: We're only checking special majorities after all terms are scanned and not when we're populating potential leaders.

File: src/kudu/consensus/leader_election.cc [hunk 4]
```diff
@@ -1062,60 +1077,59 @@ std::pair<bool, bool> FlexibleVoteCounter::IsDynamicQuorumSatisfied() const {
...
+
+    auto now = std::chrono::system_clock::now();
+    long time_elapsed_secs =
+        std::chrono::duration_cast<std::chrono::seconds>(now - creation_time_)
+            .count();
+
+    // Case: Pessimistic quorum satisfaction is possible and we still have time
+    // to wait for it
+    if (pessimistic_result.second &&
+        time_elapsed_secs < FLAGS_wait_for_pessimistic_quorum_secs) {
```
Comment (abhinav04sharma) [inline src/kudu/consensus/leader_election.cc:1099]: Overlapping with pessimistic quorum if we've not reached the timeout yet. Default timeout is 10sec which might be too much (we should probably take election timeout into account but that might make things complicated). Looking for suggestions for what to set this timeout to.

Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc:1099]: Will move this before computing from voting history to save cycles

File: src/kudu/consensus/leader_election.cc [hunk 5]
```diff
@@ -239,17 +242,15 @@ void FlexibleVoteCounter::FetchTopologyInfo() {
 }
 
 FlexibleVoteCounter::FlexibleVoteCounter(
-    const std::string& candidate_uuid,
-    int64_t election_term,
-    const LastKnownLeaderPB& last_known_leader,
-    RaftConfigPB config,
+    const std::string &candidate_uuid, int64_t election_term,
+    const LastKnownLeaderPB &last_known_leader, RaftConfigPB config,
```
Comment (yichenshen) [inline src/kudu/consensus/leader_election.cc]: Are these autoformatted by the linter? I think the default linter for meta prefers the old format.

File: src/kudu/consensus/leader_election.cc [hunk 6]
```diff
@@ -239,17 +242,15 @@ void FlexibleVoteCounter::FetchTopologyInfo() {
...
-    candidate_uuid_(candidate_uuid),
-    election_term_(election_term),
-    adjust_voter_distribution_(adjust_voter_distribution),
-    last_known_leader_(last_known_leader),
-    config_(std::move(config)) {
+    : VoteCounter(1, 1), candidate_uuid_(candidate_uuid),
+      election_term_(election_term),
+      adjust_voter_distribution_(adjust_voter_distribution),
+      last_known_leader_(last_known_leader), config_(std::move(config)) {
+  creation_time_ = std::chrono::system_clock::now();
```
Comment (yichenshen) [inline src/kudu/consensus/leader_election.cc]:
```suggestion
      last_known_leader_(last_known_leader), config_(std::move(config)),
      creation_time_(std::chrono::system_clock::now()) {
```

File: src/kudu/consensus/leader_election.cc [hunk 7]
```diff
@@ -556,32 +563,30 @@ FlexibleVoteCounter::IsMajoritySatisfiedInMajorityOfRegions() const {
...
+                      << " Satisfaction possible count: "
+                      << satisfaction_possible_count;
+  return std::make_pair<>(satisfied_count >= num_majority_regions,
+                          satisfaction_possible_count >= num_majority_regions);
 }
 
-std::pair<bool, bool>
-FlexibleVoteCounter::IsMajoritySatisfiedInPotentialLeaderRegions(
-    const std::set<std::string>& leader_regions) const {
+std::pair<bool, bool> FlexibleVoteCounter::IsMajoritySatisfiedInRegions(
```
Comment (yichenshen) [inline src/kudu/consensus/leader_election.cc]: Should we name this something else? It can be confused with the one on line 381 when reading code unless we know the input type.

File: src/kudu/consensus/leader_election.cc [hunk 8]
```diff
@@ -1003,53 +990,81 @@ std::pair<bool, bool> FlexibleVoteCounter::ComputeElectionResultFromVotingHistor
...
+  LastKnownLeaderPB last_known_leader;
+  GetLastKnownLeader(&last_known_leader);
+
   // This could be empty as a Last known leader could no more
   // be a voter or in the ring.
   std::string last_known_leader_region(
       DetermineRegionForUUID(last_known_leader.uuid()));
 
+  // Declare loss early upon discovering leader in higher term.
+  if (election_term_ <= last_known_leader.election_term()) {
```
Comment (yichenshen) [inline src/kudu/consensus/leader_election.cc:1021]: This can happen even before DetermineRegionForUUID right?

**Review by yichenshen — APPROVED** [review]
Look good besides a few nits.

File: src/kudu/consensus/leader_election.cc (hunk 4 above)
Comment (yichenshen) [reply src/kudu/consensus/leader_election.cc:1099]: Election typically finish in subsecond times, so I guess 1-2s should be more than enough.

There's also a snooze time for failure detector during the election IIRC. I think this should align with that.

File: src/kudu/consensus/leader_election.cc (hunk 5 above)
Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc]: Yes, I ran clang format (with meta configs) on the diff.

File: src/kudu/consensus/leader_election.cc (hunk 7 above)
Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc]: Let me see if I can find a better name

File: src/kudu/consensus/leader_election.cc (hunk 5 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc]: it would be nice to reduce the diff to the bare minimum. much harder to review this diff, because every line needs to be looked into. these reformats will also make taking patches from kudu down harder.

Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc]: All changes are in the flexi raft code and I even though it's a pain right now having these changes clang formatted is good in the long term since we can run clang format before every diff to be consistent. Also, I ran the clang format only on the diff changes so it only changes code around the lines that I've touched. However, if these changes are making the review very hard I can try to revert them. Let me know either way.

Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc]: If you want to do clang formatting of this file, you can use a stacked diff approach (or a separate PR). That way the no-op formatting change can be gotten out of the way with no review.

Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc]: Will do.

Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc]: Nevermind, running clang and rebasing over that is going to be a mess, I'll just remove the clang formatting from this patch manually.

**Comment by abhinav04sharma** [comment]
Updated without clang changes

File: src/kudu/consensus/leader_election.cc [hunk 9]
```diff
@@ -386,6 +391,11 @@ FlexibleVoteCounter::IsMajoritySatisfiedInRegions(
   std::vector<std::pair<bool, bool> > results;
 
   for (const std::string& region : regions) {
+    if (region.empty()) {
```
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc:394]: nit: is this check required? I think we have guarded most call sites

File: src/kudu/consensus/leader_election.cc [hunk 10]
```diff
@@ -429,8 +439,9 @@ FlexibleVoteCounter::IsMajoritySatisfiedInRegion(
   // regions and then provides quorum satisfaction information corresponding
   // to each region. Each pair of booleans represent if the quorum is already
   // satisfied and if it can be specified in a given region.
-  const std::vector<std::pair<bool, bool> >& results =
-      IsMajoritySatisfiedInRegions({region});
+  const std::vector<std::string> regions = {region};
```
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: nit: unnecessary change.

File: src/kudu/consensus/leader_election.cc [hunk 11]
```diff
@@ -507,28 +518,12 @@ FlexibleVoteCounter::IsPessimisticQuorumSatisfied() const {
   VLOG_WITH_PREFIX(3) << "Checking if pessimistic quorum is satisfied.";
 
   // Fetching all regions.
-  std::vector<std::string> regions;
+  std::set<std::string> regions;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc:520]: nit: unnecessary change. voter_distribution_ is a map. So vector insertion would have been unique.

File: src/kudu/consensus/leader_election.cc [hunk 12]
```diff
@@ -507,28 +518,12 @@ FlexibleVoteCounter::IsPessimisticQuorumSatisfied() const {
...
-  bool quorum_satisfaction_possible = true;
-
-  for (const std::pair<bool, bool>& result : results) {
-    quorum_satisfied = quorum_satisfied && result.first;
-    quorum_satisfaction_possible =
-        quorum_satisfaction_possible && result.second;
+    regions.insert(region_count_pair.first);
   }
-  return std::make_pair<>(quorum_satisfied, quorum_satisfaction_possible);
+  return IsMajoritySatisfiedInRegions(regions);
```
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: The previous code organization seems a little less confusing.
The helper function IsMajoritySatisfiedInRegions (in new code) which takes a set of 
regions and returns a single pair only makes sense when we are in Pessimistic Quorum
mode. In other modes like majority of majorities mode, we can't do simple && enforcement
on the quorum satisfaction results for each region. Therefore having it inlined in IsPessimisticQuorumSatisfied was causing less confusion.

If you want to still break it up into 2 functions for readability reasons, I would say that the
function  IsMajoritySatisfiedInRegions should be renamed as IsMajoritySatisfiedInRegionsConjunctive or IsMajoritySatisfiedInAllRegions ... something that conveys && of all regions.

File: src/kudu/consensus/leader_election.cc (hunk 9 above)
Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc:394]: I was hitting this when running unittests, might have corrected it since then but still not a bad check to have.

File: src/kudu/consensus/leader_election.cc (hunk 11 above)
Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc:520]: New IsMajoritySatisfiedInRegions() takes in a set. I'd actually change all call sites to take in a set as that makes more sense than a vector :)

File: src/kudu/consensus/leader_election.cc (hunk 12 above)
Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc]: Will change it to IsMajoritySatisfiedInAllRegions()

**Review by anirbanr-fb — COMMENTED** [review]
Can we document what the function AppendPotentialLeaderUUID is supposed to do.

**Review by anirbanr-fb — COMMENTED** [review]
Did a thorough review. The code has been refactored into more understandable
functions. Looks good to me.

A few things to do before landing.

1. Add a comment to continuity not required, that it is actually false in prod due to trust_last_leader_entries being not used
2. As a future improvement should the wait be based on number of votes received from the total number of votes expected instead of pure time based. For example we could wait for a majority of votes from all voting regions to arrive. In common case this would be much much faster. The risk with the current approach is that it bakes in the worst case wait.

File: src/kudu/consensus/leader_election.cc [hunk 13]
```diff
@@ -565,23 +559,21 @@ FlexibleVoteCounter::IsMajoritySatisfiedInMajorityOfRegions() const {
       satisfaction_possible_count >= num_majority_regions);
 }
 
-std::pair<bool, bool>
-FlexibleVoteCounter::IsMajoritySatisfiedInPotentialLeaderRegions(
-    const std::set<std::string>& leader_regions) const {
-  CHECK(!leader_regions.empty());
-  std::vector<std::string> leader_regions_vector(
-      leader_regions.begin(), leader_regions.end());
+std::pair<bool, bool> FlexibleVoteCounter::IsMajoritySatisfiedInAllRegions(
```
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc:562]: thinking about it a bit more, is "All" a bit confusing or should it be "These" ?

**Review by anirbanr-fb — APPROVED** [review]

File: src/kudu/consensus/leader_election.cc (hunk 13 above)
Comment (abhinav04sharma) [reply src/kudu/consensus/leader_election.cc:562]: I think 'All' should be fine since we're passing the regions are parameters
